{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "nbpresent": {
     "id": "122ad36c-8ca6-4194-a6c1-1c0f07903677"
    }
   },
   "source": [
    "# Introduction: Constructing a data-driven gene ontology to study disease mechanisms\n",
    "\n",
    "This notebook demonstrates how to create data-driven gene ontologies to study disease mechanisms. This notebook focuses on Fanconi Anemia (FA), a rare genetic disorder that is associated with bone marrow failure, myeloid dysplasia, and increased cancer risk. Although mutations in 20 genes are known to cause FA by dysrupting the repair of DNA damage, the existence of other FA genes and the involvement of other pathways besides DNA repair remain unclear. To discover new FA genes and pathways, this notebook executes a five-step pipeline to construct a Fanconi Anemia gene ontology (FanGO)\n",
    "\n",
    "1. Gather input data, consisting of the 20 known FA genes as a seed set of genes for modeling and a pre-computed gene similarity network derived by integrating several types of molecular evidence including protein-protein interactions, co-expression, co-localization, and epistasis.\n",
    "\n",
    "2. Score every gene for its involvement in FA by calculating its average functional similarity to the seed genes. The minimum score among the seed genes was used as a threshold to identify an additional set of 174 candidate genes.\n",
    "\n",
    "3. Organize all genes in a hierarchy of 74 cellular subsystems to construct FanGO.\n",
    "\n",
    "4. Align FanGO to the Gene Ontology.\n",
    "\n",
    "5. Upload FanGO to an online database, the Network Data Exchange ([NDEx](http://ndexbio.org)), and visualize FanGO in the [HiView](http://hiview.ucsd.edu) web application.\n",
    "\n",
    "Code is also provided to analyze 651 other diseases using a similar pipeline. Known gene-disease associations are taken from the [Monarch Initiative database](https://monarchinitiative.org/)\n",
    "\n",
    "Before reading this notebook, it is recommended that you look at the [DDOT tutorial](https://github.com/michaelkyu/ddot/blob/master/examples/Tutorial.ipynb)\n",
    "\n",
    "<img src=\"https://raw.githubusercontent.com/michaelkyu/ddot/master/docs/software_pipeline_23jan2018.png\" width=\"700\" align=\"left\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-12T05:52:04.023604Z",
     "start_time": "2018-12-12T05:52:01.870305Z"
    },
    "collapsed": true,
    "nbpresent": {
     "id": "25718bac-8f5f-4d23-a2a9-b5be1791371b"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "\n",
    "import ddot\n",
    "from ddot import Ontology"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbpresent": {
     "id": "1be29053-42b0-4eb0-bcdc-c2de670b16ce"
    }
   },
   "source": [
    "# Set username and password on the Network Data Exchange (NDEx).\n",
    "* It is strongly recommended that you create a free account on NDEx in order to keep track of your own ontologies.\n",
    "* Note that there are two NDEx servers: the main server at http://public.ndexbio.org/ and a test server for prototyping your code at http://test.ndexbio.org (also aliased as http://dev2.ndexbio.org). Each server requires a separate user account. Because the main server contains networks from publications, we recommend that you use an account on the test server while you become familiar with DDOT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-12T05:52:07.274382Z",
     "start_time": "2018-12-12T05:52:07.269637Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Note: change the server to http://public.ndexbio.org, if this is where you created your NDEx account\n",
    "ndex_server = 'http://test.ndexbio.org' \n",
    "\n",
    "# Set the NDEx server and the user account (replace with your own account)\n",
    "ndex_user, ndex_pass = '<enter your username>', '<enter your account password>'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read gene-gene integrated similarity network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Install the simplejson package (it is recommend you run this in a separate bash terminal, not in this Jupyter notebook. If you want to use a conda virtual environment, then you first need to activate the environment)\n",
    "! pip install simplejson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-12T05:59:20.520657Z",
     "start_time": "2018-12-12T05:52:09.690133Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NDEx download and CX parse time (sec): 356.2376916408539\n",
      "Dim: (19009, 19009)\n",
      "Bytes: 2890736648\n",
      "Iterate through CX and construct array time (sec): 14.60573434829712\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.654751</td>\n",
       "      <td>0.466593</td>\n",
       "      <td>0.322811</td>\n",
       "      <td>0.027936</td>\n",
       "      <td>0.451599</td>\n",
       "      <td>0.332334</td>\n",
       "      <td>0.221617</td>\n",
       "      <td>0.565551</td>\n",
       "      <td>0.563473</td>\n",
       "      <td>...</td>\n",
       "      <td>0.470512</td>\n",
       "      <td>0.093434</td>\n",
       "      <td>0.088042</td>\n",
       "      <td>0.922059</td>\n",
       "      <td>0.198680</td>\n",
       "      <td>0.991425</td>\n",
       "      <td>0.536011</td>\n",
       "      <td>0.479693</td>\n",
       "      <td>0.021175</td>\n",
       "      <td>0.290588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.654751</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.871843</td>\n",
       "      <td>0.304977</td>\n",
       "      <td>0.281618</td>\n",
       "      <td>0.465988</td>\n",
       "      <td>0.694550</td>\n",
       "      <td>0.428951</td>\n",
       "      <td>0.291614</td>\n",
       "      <td>0.970460</td>\n",
       "      <td>...</td>\n",
       "      <td>0.334991</td>\n",
       "      <td>0.475563</td>\n",
       "      <td>0.453309</td>\n",
       "      <td>0.963410</td>\n",
       "      <td>0.237216</td>\n",
       "      <td>0.781723</td>\n",
       "      <td>0.952073</td>\n",
       "      <td>0.453467</td>\n",
       "      <td>0.756576</td>\n",
       "      <td>0.301505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.466593</td>\n",
       "      <td>0.871843</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.999053</td>\n",
       "      <td>0.062145</td>\n",
       "      <td>0.820181</td>\n",
       "      <td>0.777252</td>\n",
       "      <td>0.430240</td>\n",
       "      <td>0.480298</td>\n",
       "      <td>0.816577</td>\n",
       "      <td>...</td>\n",
       "      <td>0.327862</td>\n",
       "      <td>0.455414</td>\n",
       "      <td>0.204230</td>\n",
       "      <td>0.495449</td>\n",
       "      <td>0.205282</td>\n",
       "      <td>0.255577</td>\n",
       "      <td>0.262837</td>\n",
       "      <td>0.731297</td>\n",
       "      <td>0.850221</td>\n",
       "      <td>0.379972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.322811</td>\n",
       "      <td>0.304977</td>\n",
       "      <td>0.999053</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.015770</td>\n",
       "      <td>0.474248</td>\n",
       "      <td>0.333386</td>\n",
       "      <td>0.476089</td>\n",
       "      <td>0.438315</td>\n",
       "      <td>0.346933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.189578</td>\n",
       "      <td>0.276015</td>\n",
       "      <td>0.260390</td>\n",
       "      <td>0.968592</td>\n",
       "      <td>0.047822</td>\n",
       "      <td>0.890704</td>\n",
       "      <td>0.282092</td>\n",
       "      <td>0.447627</td>\n",
       "      <td>0.663615</td>\n",
       "      <td>0.100694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>0.027936</td>\n",
       "      <td>0.281618</td>\n",
       "      <td>0.062145</td>\n",
       "      <td>0.015770</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.049229</td>\n",
       "      <td>0.009654</td>\n",
       "      <td>0.083807</td>\n",
       "      <td>0.155461</td>\n",
       "      <td>0.193261</td>\n",
       "      <td>...</td>\n",
       "      <td>0.318918</td>\n",
       "      <td>0.377512</td>\n",
       "      <td>0.259443</td>\n",
       "      <td>0.514783</td>\n",
       "      <td>0.197062</td>\n",
       "      <td>0.574311</td>\n",
       "      <td>0.825521</td>\n",
       "      <td>0.335596</td>\n",
       "      <td>0.608165</td>\n",
       "      <td>0.063526</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19009 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             A1BG      A1CF       A2M     A2ML1   A3GALT2    A4GALT     A4GNT  \\\n",
       "A1BG     0.000053  0.654751  0.466593  0.322811  0.027936  0.451599  0.332334   \n",
       "A1CF     0.654751  0.000053  0.871843  0.304977  0.281618  0.465988  0.694550   \n",
       "A2M      0.466593  0.871843  0.000053  0.999053  0.062145  0.820181  0.777252   \n",
       "A2ML1    0.322811  0.304977  0.999053  0.000053  0.015770  0.474248  0.333386   \n",
       "A3GALT2  0.027936  0.281618  0.062145  0.015770  0.000053  0.049229  0.009654   \n",
       "\n",
       "             AAAS      AACS     AADAC    ...       ZWILCH     ZWINT      ZXDA  \\\n",
       "A1BG     0.221617  0.565551  0.563473    ...     0.470512  0.093434  0.088042   \n",
       "A1CF     0.428951  0.291614  0.970460    ...     0.334991  0.475563  0.453309   \n",
       "A2M      0.430240  0.480298  0.816577    ...     0.327862  0.455414  0.204230   \n",
       "A2ML1    0.476089  0.438315  0.346933    ...     0.189578  0.276015  0.260390   \n",
       "A3GALT2  0.083807  0.155461  0.193261    ...     0.318918  0.377512  0.259443   \n",
       "\n",
       "             ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
       "A1BG     0.922059  0.198680  0.991425  0.536011  0.479693  0.021175  0.290588  \n",
       "A1CF     0.963410  0.237216  0.781723  0.952073  0.453467  0.756576  0.301505  \n",
       "A2M      0.495449  0.205282  0.255577  0.262837  0.731297  0.850221  0.379972  \n",
       "A2ML1    0.968592  0.047822  0.890704  0.282092  0.447627  0.663615  0.100694  \n",
       "A3GALT2  0.514783  0.197062  0.574311  0.825521  0.335596  0.608165  0.063526  \n",
       "\n",
       "[5 rows x 19009 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Download human gene-gene similarity network from NDEx\n",
    "## -- WARNING: This network is very large (19,009-by-19,009 matrix).\n",
    "## -- *** Requires ~6 GB of RAM to download and process ***\n",
    "## -- *** ~4 GB of data is downloaded from NDEx, which takes ~10 min for a fast internet connection ***\n",
    "## -- This requires the simplejson package (see previous cell)\n",
    "sim, sim_names = ddot.ndex_to_sim_matrix(\n",
    "    ndex_url=ddot.HUMAN_GENE_SIMILARITIES_URL,\n",
    "    input_fmt='cx_matrix',\n",
    "    output_fmt='matrix')\n",
    "sim = pd.DataFrame(sim, columns=sim_names, index=sim_names)\n",
    "\n",
    "## Rank transform the similarities\n",
    "sim_rank = sim.rank(0) / (sim.shape[0] - 1)\n",
    "sim_rank = pd.DataFrame((sim_rank.values + sim_rank.values.T) / 2.0, columns=sim_names, index=sim_names)\n",
    "\n",
    "sim_rank.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbpresent": {
     "id": "f0d07107-fd03-42b8-bf19-fdf2ffde5b61"
    }
   },
   "source": [
    "# Specify a set of seed genes with known associations to the disease being studied"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:24:58.268288Z",
     "start_time": "2018-12-07T22:24:58.265012Z"
    },
    "collapsed": true,
    "nbpresent": {
     "id": "18901780-c455-4bac-a323-4d4a146c4de1"
    }
   },
   "outputs": [],
   "source": [
    "# Let seed genes be the 20 known genes that cause Fanconi Anemia (from the Fanconi Anemia Mutation Database, http://www2.rockefeller.edu/fanconi/)\n",
    "seed = ['FANCA', 'FANCB', 'FANCC', 'BRCA2', 'FANCD2',\n",
    "        'FANCE', 'FANCF', 'FANCG', 'FANCI', 'BRIP1',\n",
    "        'FANCL', 'FANCM', 'PALB2', 'RAD51C', 'SLX4',\n",
    "        'ERCC4', 'RAD51', 'BRCA1', 'UBE2T', 'XRCC2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-08T19:48:55.626301Z",
     "start_time": "2018-06-08T19:48:55.621255Z"
    },
    "collapsed": true,
    "nbpresent": {
     "id": "628f6bcc-91c6-4807-90cb-2f82b12a2892"
    }
   },
   "outputs": [],
   "source": [
    "# # Let seed genes be the known genese for one of 651 diseases (uncomment to use)\n",
    "\n",
    "# # Retrieve a table of gene-disease associations from the Monarch Initiative (reformatted and stored on NDEx)\n",
    "# monarch, _ = ddot.ndex_to_sim_matrix(\n",
    "#     ddot.config.MONARCH_DISEASE_GENE_SLIM_URL,\n",
    "#     input_fmt='cx',\n",
    "#     output_fmt='sparse')\n",
    "# print(monarch.head())\n",
    "\n",
    "# # Example: get the known genes for Caffey Disease\n",
    "# seed = monarch.loc[monarch['disease']=='caffey_disease', 'gene'].tolist()\n",
    "# seed = [s for s in seed if s in sim_names]\n",
    "# print('Seed:', seed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbpresent": {
     "id": "f0261c45-a5fc-443a-989b-90e07507d20e"
    }
   },
   "source": [
    "# Identify candidate set of genes that are highly similar to the seed set of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:26:30.821146Z",
     "start_time": "2018-12-07T22:26:24.716688Z"
    },
    "collapsed": true,
    "nbpresent": {
     "id": "38b31ac6-24d7-4012-9497-2e1114d548d2"
    }
   },
   "outputs": [],
   "source": [
    "expand, expand_idx, sim_2_seed, fig = ddot.expand_seed(\n",
    "    seed,\n",
    "    sim_rank.values,\n",
    "    sim_names,\n",
    "    seed_perc=0,\n",
    "    agg='mean',\n",
    "    figure=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbpresent": {
     "id": "aa21f852-ff5b-4380-82c9-0caa92162ead"
    }
   },
   "source": [
    "# Organize seed and candidate genes into a data-driven gene ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:26:31.541840Z",
     "start_time": "2018-12-07T22:26:30.825155Z"
    },
    "nbpresent": {
     "id": "080d0cb5-6d63-44e2-8d94-165a2cd7ad81"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "194 genes, 74 terms, 349 gene-term relations, 86 term-term relations\n",
       "node_attributes: ['Parent weight']\n",
       "edge_attributes: ['CLIXO_score']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run CliXO, with parameters alpha=0.05 and beta=0.5\n",
    "ont = Ontology.run_clixo(sim.loc[expand, :].loc[:, expand], alpha=0.05, beta=0.5, square=True)\n",
    "ont"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbpresent": {
     "id": "260b3d19-9afc-4fe4-91b1-7a3fbd3c1135"
    }
   },
   "source": [
    "# Align the data-driven ontology with the Gene Ontology (GO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:27:04.448917Z",
     "start_time": "2018-12-07T22:26:31.545331Z"
    },
    "nbpresent": {
     "id": "8d454a74-6744-48ec-ab46-b8565c860512"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "19015 genes, 19343 terms, 215488 gene-term relations, 36362 term-term relations\n",
      "node_attributes: ['Term_Description', 'name', 'Branch', 'Vis:Fill Color', 'Vis:Border Paint', 'Vis:Shape']\n",
      "edge_attributes: ['Vis:Visible']\n"
     ]
    }
   ],
   "source": [
    "# Read Gene Ontology from NDEx. This version has been pre-processed to contain a non-redundant set of GO terms and connections that are relevant to human genes (see Get_Gene_Ontology.ipynb) \n",
    "go_human = Ontology.from_ndex(ddot.config.GO_HUMAN_URL)\n",
    "print(go_human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:28:09.001224Z",
     "start_time": "2018-12-07T22:28:01.705322Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common genes: 193\n",
      "collapse command: /cellar/users/mikeyu/anaconda2/envs/ddot_py36/lib/python3.6/site-packages/ddot/alignOntology/collapseRedundantNodes /tmp/tmpur4okjlk\n",
      "collapse command: /cellar/users/mikeyu/anaconda2/envs/ddot_py36/lib/python3.6/site-packages/ddot/alignOntology/collapseRedundantNodes /tmp/tmpibt10j5g\n",
      "ont1_collapsed: 193 genes, 74 terms, 348 gene-term relations, 86 term-term relations\n",
      "node_attributes: ['Parent weight']\n",
      "edge_attributes: ['CLIXO_score']\n",
      "ont2_collapsed: 193 genes, 1854 terms, 3594 gene-term relations, 4490 term-term relations\n",
      "node_attributes: ['Term_Description', 'name', 'Branch', 'Vis:Fill Color', 'Vis:Border Paint', 'Vis:Shape']\n",
      "edge_attributes: ['Vis:Visible']\n",
      "Alignment command: /cellar/users/mikeyu/anaconda2/envs/ddot_py36/lib/python3.6/site-packages/ddot/alignOntology/calculateFDRs /tmp/tmpk0eckbod /tmp/tmprb3m4od4 0.05 criss_cross /tmp/tmpczovc09t 100 40 gene\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Similarity</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S:196</th>\n",
       "      <td>GO:0035098</td>\n",
       "      <td>0.902149</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:219</th>\n",
       "      <td>GO:1905773</td>\n",
       "      <td>0.900885</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:206</th>\n",
       "      <td>GO:1901673</td>\n",
       "      <td>0.891945</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:205</th>\n",
       "      <td>GO:0004748</td>\n",
       "      <td>0.891813</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:240</th>\n",
       "      <td>GO:0071821</td>\n",
       "      <td>0.891386</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Term  Similarity  FDR\n",
       "Term                              \n",
       "S:196  GO:0035098    0.902149  0.0\n",
       "S:219  GO:1905773    0.900885  0.0\n",
       "S:206  GO:1901673    0.891945  0.0\n",
       "S:205  GO:0004748    0.891813  0.0\n",
       "S:240  GO:0071821    0.891386  0.0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Align ontologies\n",
    "alignment = ont.align(go_human, \n",
    "                      iterations=100,\n",
    "                      update_self=['Term_Description'],\n",
    "                      align_label='Term_Description',\n",
    "                      verbose=True)\n",
    "alignment.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:28:09.035813Z",
     "start_time": "2018-12-07T22:28:09.003667Z"
    },
    "nbpresent": {
     "id": "ead00b01-64fe-439c-a274-15f303a3d5a4"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent weight</th>\n",
       "      <th>Aligned_Term</th>\n",
       "      <th>Aligned_Similarity</th>\n",
       "      <th>Aligned_FDR</th>\n",
       "      <th>Aligned_Term_Description</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S:194</th>\n",
       "      <td>7.90340</td>\n",
       "      <td>GO:0005873</td>\n",
       "      <td>0.441441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>plus-end kinesin complex</td>\n",
       "      <td>S:194\\nplus-end kinesin complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:195</th>\n",
       "      <td>8.22255</td>\n",
       "      <td>GO:0098535</td>\n",
       "      <td>0.889600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>de novo centriole assembly involved in multi-c...</td>\n",
       "      <td>S:195\\nde novo centriole assembly involved in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:196</th>\n",
       "      <td>8.18291</td>\n",
       "      <td>GO:0035098</td>\n",
       "      <td>0.902149</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ESC/E(Z) complex</td>\n",
       "      <td>S:196\\nESC/E(Z) complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:197</th>\n",
       "      <td>8.13613</td>\n",
       "      <td>GO:0001882</td>\n",
       "      <td>0.450824</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nucleoside binding</td>\n",
       "      <td>S:197\\nnucleoside binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:198</th>\n",
       "      <td>8.13260</td>\n",
       "      <td>GO:0000942</td>\n",
       "      <td>0.300136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>condensed nuclear chromosome outer kinetochore</td>\n",
       "      <td>S:198\\ncondensed nuclear chromosome outer kine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:199</th>\n",
       "      <td>8.09260</td>\n",
       "      <td>GO:0000732</td>\n",
       "      <td>0.101758</td>\n",
       "      <td>0.0</td>\n",
       "      <td>strand displacement</td>\n",
       "      <td>S:199\\nstrand displacement</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:200</th>\n",
       "      <td>7.99652</td>\n",
       "      <td>GO:0051087</td>\n",
       "      <td>0.505705</td>\n",
       "      <td>0.0</td>\n",
       "      <td>chaperone binding</td>\n",
       "      <td>S:200\\nchaperone binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:201</th>\n",
       "      <td>7.99062</td>\n",
       "      <td>GO:0097472</td>\n",
       "      <td>0.447222</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cyclin-dependent protein kinase activity</td>\n",
       "      <td>S:201\\ncyclin-dependent protein kinase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:202</th>\n",
       "      <td>7.98245</td>\n",
       "      <td>GO:0000320</td>\n",
       "      <td>0.462717</td>\n",
       "      <td>0.0</td>\n",
       "      <td>re-entry into mitotic cell cycle</td>\n",
       "      <td>S:202\\nre-entry into mitotic cell cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:203</th>\n",
       "      <td>7.96387</td>\n",
       "      <td>GO:1990414</td>\n",
       "      <td>0.447235</td>\n",
       "      <td>0.0</td>\n",
       "      <td>replication-born double-strand break repair vi...</td>\n",
       "      <td>S:203\\nreplication-born double-strand break re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:204</th>\n",
       "      <td>7.96025</td>\n",
       "      <td>GO:1990086</td>\n",
       "      <td>0.446096</td>\n",
       "      <td>0.0</td>\n",
       "      <td>lens fiber cell apoptotic process</td>\n",
       "      <td>S:204\\nlens fiber cell apoptotic process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:205</th>\n",
       "      <td>7.91668</td>\n",
       "      <td>GO:0004748</td>\n",
       "      <td>0.891813</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ribonucleoside-diphosphate reductase activity,...</td>\n",
       "      <td>S:205\\nribonucleoside-diphosphate reductase ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:206</th>\n",
       "      <td>9.13645</td>\n",
       "      <td>GO:1901673</td>\n",
       "      <td>0.891945</td>\n",
       "      <td>0.0</td>\n",
       "      <td>regulation of mitotic spindle assembly</td>\n",
       "      <td>S:206\\nregulation of mitotic spindle assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:207</th>\n",
       "      <td>7.85010</td>\n",
       "      <td>GO:0000381</td>\n",
       "      <td>0.882179</td>\n",
       "      <td>0.0</td>\n",
       "      <td>regulation of alternative mRNA splicing, via s...</td>\n",
       "      <td>S:207\\nregulation of alternative mRNA splicing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:208</th>\n",
       "      <td>7.84492</td>\n",
       "      <td>GO:0070063</td>\n",
       "      <td>0.452919</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RNA polymerase binding</td>\n",
       "      <td>S:208\\nRNA polymerase binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:209</th>\n",
       "      <td>7.81645</td>\n",
       "      <td>GO:0008536</td>\n",
       "      <td>0.598753</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Ran GTPase binding</td>\n",
       "      <td>S:209\\nRan GTPase binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:210</th>\n",
       "      <td>7.78127</td>\n",
       "      <td>GO:0051292</td>\n",
       "      <td>0.883572</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nuclear pore complex assembly</td>\n",
       "      <td>S:210\\nnuclear pore complex assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:211</th>\n",
       "      <td>7.77098</td>\n",
       "      <td>GO:0006346</td>\n",
       "      <td>0.868901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>methylation-dependent chromatin silencing</td>\n",
       "      <td>S:211\\nmethylation-dependent chromatin silencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:212</th>\n",
       "      <td>7.22840</td>\n",
       "      <td>GO:0070064</td>\n",
       "      <td>0.514149</td>\n",
       "      <td>0.0</td>\n",
       "      <td>proline-rich region binding</td>\n",
       "      <td>S:212\\nproline-rich region binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:213</th>\n",
       "      <td>7.13854</td>\n",
       "      <td>GO:0048025</td>\n",
       "      <td>0.878031</td>\n",
       "      <td>0.0</td>\n",
       "      <td>negative regulation of mRNA splicing, via spli...</td>\n",
       "      <td>S:213\\nnegative regulation of mRNA splicing, v...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:214</th>\n",
       "      <td>7.07749</td>\n",
       "      <td>GO:0009113</td>\n",
       "      <td>0.882894</td>\n",
       "      <td>0.0</td>\n",
       "      <td>purine nucleobase biosynthetic process</td>\n",
       "      <td>S:214\\npurine nucleobase biosynthetic process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:215</th>\n",
       "      <td>6.88769</td>\n",
       "      <td>GO:0002020</td>\n",
       "      <td>0.308064</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protease binding</td>\n",
       "      <td>S:215\\nprotease binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:216</th>\n",
       "      <td>6.87976</td>\n",
       "      <td>GO:0032042</td>\n",
       "      <td>0.573594</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mitochondrial DNA metabolic process</td>\n",
       "      <td>S:216\\nmitochondrial DNA metabolic process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:217</th>\n",
       "      <td>8.49673</td>\n",
       "      <td>GO:0043137</td>\n",
       "      <td>0.586344</td>\n",
       "      <td>0.0</td>\n",
       "      <td>DNA replication, removal of RNA primer</td>\n",
       "      <td>S:217\\nDNA replication, removal of RNA primer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:218</th>\n",
       "      <td>8.66391</td>\n",
       "      <td>GO:0007089</td>\n",
       "      <td>0.446753</td>\n",
       "      <td>0.0</td>\n",
       "      <td>traversing start control point of mitotic cell...</td>\n",
       "      <td>S:218\\ntraversing start control point of mitot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:219</th>\n",
       "      <td>8.64544</td>\n",
       "      <td>GO:1905773</td>\n",
       "      <td>0.900885</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8-hydroxy-2'-deoxyguanosine DNA binding</td>\n",
       "      <td>S:219\\n8-hydroxy-2'-deoxyguanosine DNA binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:220</th>\n",
       "      <td>8.68338</td>\n",
       "      <td>GO:0032301</td>\n",
       "      <td>0.887632</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MutSalpha complex</td>\n",
       "      <td>S:220\\nMutSalpha complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:221</th>\n",
       "      <td>8.58473</td>\n",
       "      <td>GO:0035563</td>\n",
       "      <td>0.878671</td>\n",
       "      <td>0.0</td>\n",
       "      <td>positive regulation of chromatin binding</td>\n",
       "      <td>S:221\\npositive regulation of chromatin binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:222</th>\n",
       "      <td>8.36104</td>\n",
       "      <td>GO:0042177</td>\n",
       "      <td>0.445886</td>\n",
       "      <td>0.0</td>\n",
       "      <td>negative regulation of protein catabolic process</td>\n",
       "      <td>S:222\\nnegative regulation of protein cataboli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:223</th>\n",
       "      <td>9.14107</td>\n",
       "      <td>GO:0070530</td>\n",
       "      <td>0.446841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>K63-linked polyubiquitin modification-dependen...</td>\n",
       "      <td>S:223\\nK63-linked polyubiquitin modification-d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:238</th>\n",
       "      <td>7.39080</td>\n",
       "      <td>GO:0001741</td>\n",
       "      <td>0.337589</td>\n",
       "      <td>0.0</td>\n",
       "      <td>XY body</td>\n",
       "      <td>S:238\\nXY body</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:239</th>\n",
       "      <td>8.34847</td>\n",
       "      <td>GO:0000811</td>\n",
       "      <td>0.600984</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GINS complex</td>\n",
       "      <td>S:239\\nGINS complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:240</th>\n",
       "      <td>8.23992</td>\n",
       "      <td>GO:0071821</td>\n",
       "      <td>0.891386</td>\n",
       "      <td>0.0</td>\n",
       "      <td>FANCM-MHF complex</td>\n",
       "      <td>S:240\\nFANCM-MHF complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:241</th>\n",
       "      <td>7.74013</td>\n",
       "      <td>GO:0098536</td>\n",
       "      <td>0.432697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>deuterosome</td>\n",
       "      <td>S:241\\ndeuterosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:242</th>\n",
       "      <td>7.85766</td>\n",
       "      <td>GO:0006335</td>\n",
       "      <td>0.360899</td>\n",
       "      <td>0.0</td>\n",
       "      <td>DNA replication-dependent nucleosome assembly</td>\n",
       "      <td>S:242\\nDNA replication-dependent nucleosome as...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:243</th>\n",
       "      <td>8.66652</td>\n",
       "      <td>GO:0008821</td>\n",
       "      <td>0.592757</td>\n",
       "      <td>0.0</td>\n",
       "      <td>crossover junction endodeoxyribonuclease activity</td>\n",
       "      <td>S:243\\ncrossover junction endodeoxyribonucleas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:244</th>\n",
       "      <td>8.39262</td>\n",
       "      <td>GO:0008610</td>\n",
       "      <td>0.442373</td>\n",
       "      <td>0.0</td>\n",
       "      <td>lipid biosynthetic process</td>\n",
       "      <td>S:244\\nlipid biosynthetic process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:245</th>\n",
       "      <td>8.53910</td>\n",
       "      <td>GO:0000150</td>\n",
       "      <td>0.552376</td>\n",
       "      <td>0.0</td>\n",
       "      <td>recombinase activity</td>\n",
       "      <td>S:245\\nrecombinase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:246</th>\n",
       "      <td>8.13655</td>\n",
       "      <td>GO:0005664</td>\n",
       "      <td>0.251620</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nuclear origin of replication recognition complex</td>\n",
       "      <td>S:246\\nnuclear origin of replication recogniti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:247</th>\n",
       "      <td>8.10275</td>\n",
       "      <td>GO:0043240</td>\n",
       "      <td>0.414114</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Fanconi anaemia nuclear complex</td>\n",
       "      <td>S:247\\nFanconi anaemia nuclear complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:248</th>\n",
       "      <td>8.64773</td>\n",
       "      <td>GO:0006978</td>\n",
       "      <td>0.350815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>DNA damage response, signal transduction by p5...</td>\n",
       "      <td>S:248\\nDNA damage response, signal transductio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:249</th>\n",
       "      <td>7.50253</td>\n",
       "      <td>GO:0031261</td>\n",
       "      <td>0.371194</td>\n",
       "      <td>0.0</td>\n",
       "      <td>DNA replication preinitiation complex</td>\n",
       "      <td>S:249\\nDNA replication preinitiation complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:250</th>\n",
       "      <td>7.33911</td>\n",
       "      <td>GO:0031390</td>\n",
       "      <td>0.499434</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Ctf18 RFC-like complex</td>\n",
       "      <td>S:250\\nCtf18 RFC-like complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:251</th>\n",
       "      <td>7.58754</td>\n",
       "      <td>GO:0000712</td>\n",
       "      <td>0.402787</td>\n",
       "      <td>0.0</td>\n",
       "      <td>resolution of meiotic recombination intermediates</td>\n",
       "      <td>S:251\\nresolution of meiotic recombination int...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:252</th>\n",
       "      <td>7.79665</td>\n",
       "      <td>GO:0000796</td>\n",
       "      <td>0.448329</td>\n",
       "      <td>0.0</td>\n",
       "      <td>condensin complex</td>\n",
       "      <td>S:252\\ncondensin complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:253</th>\n",
       "      <td>7.67372</td>\n",
       "      <td>GO:0032993</td>\n",
       "      <td>0.208859</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protein-DNA complex</td>\n",
       "      <td>S:253\\nprotein-DNA complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:254</th>\n",
       "      <td>7.11545</td>\n",
       "      <td>GO:0032201</td>\n",
       "      <td>0.394536</td>\n",
       "      <td>0.0</td>\n",
       "      <td>telomere maintenance via semi-conservative rep...</td>\n",
       "      <td>S:254\\ntelomere maintenance via semi-conservat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:255</th>\n",
       "      <td>6.97441</td>\n",
       "      <td>GO:0042393</td>\n",
       "      <td>0.242748</td>\n",
       "      <td>0.0</td>\n",
       "      <td>histone binding</td>\n",
       "      <td>S:255\\nhistone binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:256</th>\n",
       "      <td>7.54420</td>\n",
       "      <td>GO:0000793</td>\n",
       "      <td>0.283442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>condensed chromosome</td>\n",
       "      <td>S:256\\ncondensed chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:257</th>\n",
       "      <td>6.71049</td>\n",
       "      <td>GO:1901363</td>\n",
       "      <td>0.272814</td>\n",
       "      <td>0.0</td>\n",
       "      <td>heterocyclic compound binding</td>\n",
       "      <td>S:257\\nheterocyclic compound binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:258</th>\n",
       "      <td>7.17072</td>\n",
       "      <td>GO:0051301</td>\n",
       "      <td>0.410023</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell division</td>\n",
       "      <td>S:258\\ncell division</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:259</th>\n",
       "      <td>8.01514</td>\n",
       "      <td>GO:0006281</td>\n",
       "      <td>0.381898</td>\n",
       "      <td>0.0</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>S:259\\nDNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:260</th>\n",
       "      <td>7.82196</td>\n",
       "      <td>GO:0006974</td>\n",
       "      <td>0.525971</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "      <td>S:260\\ncellular response to DNA damage stimulus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:261</th>\n",
       "      <td>7.38362</td>\n",
       "      <td>GO:0007049</td>\n",
       "      <td>0.591802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cell cycle</td>\n",
       "      <td>S:261\\ncell cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:262</th>\n",
       "      <td>6.91500</td>\n",
       "      <td>GO:0044446</td>\n",
       "      <td>0.688551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>intracellular organelle part</td>\n",
       "      <td>S:262\\nintracellular organelle part</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:263</th>\n",
       "      <td>6.69429</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>0.732987</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>S:263\\nnucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:264</th>\n",
       "      <td>6.38314</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>0.760222</td>\n",
       "      <td>0.0</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>S:264\\nprotein binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:265</th>\n",
       "      <td>6.21823</td>\n",
       "      <td>GO:0005488</td>\n",
       "      <td>0.797567</td>\n",
       "      <td>0.0</td>\n",
       "      <td>binding</td>\n",
       "      <td>S:265\\nbinding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:266</th>\n",
       "      <td>5.54631</td>\n",
       "      <td>GO:0009987</td>\n",
       "      <td>0.813078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>cellular process</td>\n",
       "      <td>S:266\\ncellular process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S:267</th>\n",
       "      <td>4.28603</td>\n",
       "      <td>GO:0044424</td>\n",
       "      <td>0.759738</td>\n",
       "      <td>0.0</td>\n",
       "      <td>intracellular part</td>\n",
       "      <td>S:267\\nintracellular part</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Parent weight Aligned_Term  Aligned_Similarity  Aligned_FDR  \\\n",
       "S:194        7.90340   GO:0005873            0.441441          0.0   \n",
       "S:195        8.22255   GO:0098535            0.889600          0.0   \n",
       "S:196        8.18291   GO:0035098            0.902149          0.0   \n",
       "S:197        8.13613   GO:0001882            0.450824          0.0   \n",
       "S:198        8.13260   GO:0000942            0.300136          0.0   \n",
       "S:199        8.09260   GO:0000732            0.101758          0.0   \n",
       "S:200        7.99652   GO:0051087            0.505705          0.0   \n",
       "S:201        7.99062   GO:0097472            0.447222          0.0   \n",
       "S:202        7.98245   GO:0000320            0.462717          0.0   \n",
       "S:203        7.96387   GO:1990414            0.447235          0.0   \n",
       "S:204        7.96025   GO:1990086            0.446096          0.0   \n",
       "S:205        7.91668   GO:0004748            0.891813          0.0   \n",
       "S:206        9.13645   GO:1901673            0.891945          0.0   \n",
       "S:207        7.85010   GO:0000381            0.882179          0.0   \n",
       "S:208        7.84492   GO:0070063            0.452919          0.0   \n",
       "S:209        7.81645   GO:0008536            0.598753          0.0   \n",
       "S:210        7.78127   GO:0051292            0.883572          0.0   \n",
       "S:211        7.77098   GO:0006346            0.868901          0.0   \n",
       "S:212        7.22840   GO:0070064            0.514149          0.0   \n",
       "S:213        7.13854   GO:0048025            0.878031          0.0   \n",
       "S:214        7.07749   GO:0009113            0.882894          0.0   \n",
       "S:215        6.88769   GO:0002020            0.308064          0.0   \n",
       "S:216        6.87976   GO:0032042            0.573594          0.0   \n",
       "S:217        8.49673   GO:0043137            0.586344          0.0   \n",
       "S:218        8.66391   GO:0007089            0.446753          0.0   \n",
       "S:219        8.64544   GO:1905773            0.900885          0.0   \n",
       "S:220        8.68338   GO:0032301            0.887632          0.0   \n",
       "S:221        8.58473   GO:0035563            0.878671          0.0   \n",
       "S:222        8.36104   GO:0042177            0.445886          0.0   \n",
       "S:223        9.14107   GO:0070530            0.446841          0.0   \n",
       "...              ...          ...                 ...          ...   \n",
       "S:238        7.39080   GO:0001741            0.337589          0.0   \n",
       "S:239        8.34847   GO:0000811            0.600984          0.0   \n",
       "S:240        8.23992   GO:0071821            0.891386          0.0   \n",
       "S:241        7.74013   GO:0098536            0.432697          0.0   \n",
       "S:242        7.85766   GO:0006335            0.360899          0.0   \n",
       "S:243        8.66652   GO:0008821            0.592757          0.0   \n",
       "S:244        8.39262   GO:0008610            0.442373          0.0   \n",
       "S:245        8.53910   GO:0000150            0.552376          0.0   \n",
       "S:246        8.13655   GO:0005664            0.251620          0.0   \n",
       "S:247        8.10275   GO:0043240            0.414114          0.0   \n",
       "S:248        8.64773   GO:0006978            0.350815          0.0   \n",
       "S:249        7.50253   GO:0031261            0.371194          0.0   \n",
       "S:250        7.33911   GO:0031390            0.499434          0.0   \n",
       "S:251        7.58754   GO:0000712            0.402787          0.0   \n",
       "S:252        7.79665   GO:0000796            0.448329          0.0   \n",
       "S:253        7.67372   GO:0032993            0.208859          0.0   \n",
       "S:254        7.11545   GO:0032201            0.394536          0.0   \n",
       "S:255        6.97441   GO:0042393            0.242748          0.0   \n",
       "S:256        7.54420   GO:0000793            0.283442          0.0   \n",
       "S:257        6.71049   GO:1901363            0.272814          0.0   \n",
       "S:258        7.17072   GO:0051301            0.410023          0.0   \n",
       "S:259        8.01514   GO:0006281            0.381898          0.0   \n",
       "S:260        7.82196   GO:0006974            0.525971          0.0   \n",
       "S:261        7.38362   GO:0007049            0.591802          0.0   \n",
       "S:262        6.91500   GO:0044446            0.688551          0.0   \n",
       "S:263        6.69429   GO:0005634            0.732987          0.0   \n",
       "S:264        6.38314   GO:0005515            0.760222          0.0   \n",
       "S:265        6.21823   GO:0005488            0.797567          0.0   \n",
       "S:266        5.54631   GO:0009987            0.813078          0.0   \n",
       "S:267        4.28603   GO:0044424            0.759738          0.0   \n",
       "\n",
       "                                Aligned_Term_Description  \\\n",
       "S:194                           plus-end kinesin complex   \n",
       "S:195  de novo centriole assembly involved in multi-c...   \n",
       "S:196                                   ESC/E(Z) complex   \n",
       "S:197                                 nucleoside binding   \n",
       "S:198     condensed nuclear chromosome outer kinetochore   \n",
       "S:199                                strand displacement   \n",
       "S:200                                  chaperone binding   \n",
       "S:201           cyclin-dependent protein kinase activity   \n",
       "S:202                   re-entry into mitotic cell cycle   \n",
       "S:203  replication-born double-strand break repair vi...   \n",
       "S:204                  lens fiber cell apoptotic process   \n",
       "S:205  ribonucleoside-diphosphate reductase activity,...   \n",
       "S:206             regulation of mitotic spindle assembly   \n",
       "S:207  regulation of alternative mRNA splicing, via s...   \n",
       "S:208                             RNA polymerase binding   \n",
       "S:209                                 Ran GTPase binding   \n",
       "S:210                      nuclear pore complex assembly   \n",
       "S:211          methylation-dependent chromatin silencing   \n",
       "S:212                        proline-rich region binding   \n",
       "S:213  negative regulation of mRNA splicing, via spli...   \n",
       "S:214             purine nucleobase biosynthetic process   \n",
       "S:215                                   protease binding   \n",
       "S:216                mitochondrial DNA metabolic process   \n",
       "S:217             DNA replication, removal of RNA primer   \n",
       "S:218  traversing start control point of mitotic cell...   \n",
       "S:219            8-hydroxy-2'-deoxyguanosine DNA binding   \n",
       "S:220                                  MutSalpha complex   \n",
       "S:221           positive regulation of chromatin binding   \n",
       "S:222   negative regulation of protein catabolic process   \n",
       "S:223  K63-linked polyubiquitin modification-dependen...   \n",
       "...                                                  ...   \n",
       "S:238                                            XY body   \n",
       "S:239                                       GINS complex   \n",
       "S:240                                  FANCM-MHF complex   \n",
       "S:241                                        deuterosome   \n",
       "S:242      DNA replication-dependent nucleosome assembly   \n",
       "S:243  crossover junction endodeoxyribonuclease activity   \n",
       "S:244                         lipid biosynthetic process   \n",
       "S:245                               recombinase activity   \n",
       "S:246  nuclear origin of replication recognition complex   \n",
       "S:247                    Fanconi anaemia nuclear complex   \n",
       "S:248  DNA damage response, signal transduction by p5...   \n",
       "S:249              DNA replication preinitiation complex   \n",
       "S:250                             Ctf18 RFC-like complex   \n",
       "S:251  resolution of meiotic recombination intermediates   \n",
       "S:252                                  condensin complex   \n",
       "S:253                                protein-DNA complex   \n",
       "S:254  telomere maintenance via semi-conservative rep...   \n",
       "S:255                                    histone binding   \n",
       "S:256                               condensed chromosome   \n",
       "S:257                      heterocyclic compound binding   \n",
       "S:258                                      cell division   \n",
       "S:259                                         DNA repair   \n",
       "S:260           cellular response to DNA damage stimulus   \n",
       "S:261                                         cell cycle   \n",
       "S:262                       intracellular organelle part   \n",
       "S:263                                            nucleus   \n",
       "S:264                                    protein binding   \n",
       "S:265                                            binding   \n",
       "S:266                                   cellular process   \n",
       "S:267                                 intracellular part   \n",
       "\n",
       "                                                   Label  \n",
       "S:194                    S:194\\nplus-end kinesin complex  \n",
       "S:195  S:195\\nde novo centriole assembly involved in ...  \n",
       "S:196                            S:196\\nESC/E(Z) complex  \n",
       "S:197                          S:197\\nnucleoside binding  \n",
       "S:198  S:198\\ncondensed nuclear chromosome outer kine...  \n",
       "S:199                         S:199\\nstrand displacement  \n",
       "S:200                           S:200\\nchaperone binding  \n",
       "S:201    S:201\\ncyclin-dependent protein kinase activity  \n",
       "S:202            S:202\\nre-entry into mitotic cell cycle  \n",
       "S:203  S:203\\nreplication-born double-strand break re...  \n",
       "S:204           S:204\\nlens fiber cell apoptotic process  \n",
       "S:205  S:205\\nribonucleoside-diphosphate reductase ac...  \n",
       "S:206      S:206\\nregulation of mitotic spindle assembly  \n",
       "S:207  S:207\\nregulation of alternative mRNA splicing...  \n",
       "S:208                      S:208\\nRNA polymerase binding  \n",
       "S:209                          S:209\\nRan GTPase binding  \n",
       "S:210               S:210\\nnuclear pore complex assembly  \n",
       "S:211   S:211\\nmethylation-dependent chromatin silencing  \n",
       "S:212                 S:212\\nproline-rich region binding  \n",
       "S:213  S:213\\nnegative regulation of mRNA splicing, v...  \n",
       "S:214      S:214\\npurine nucleobase biosynthetic process  \n",
       "S:215                            S:215\\nprotease binding  \n",
       "S:216         S:216\\nmitochondrial DNA metabolic process  \n",
       "S:217      S:217\\nDNA replication, removal of RNA primer  \n",
       "S:218  S:218\\ntraversing start control point of mitot...  \n",
       "S:219     S:219\\n8-hydroxy-2'-deoxyguanosine DNA binding  \n",
       "S:220                           S:220\\nMutSalpha complex  \n",
       "S:221    S:221\\npositive regulation of chromatin binding  \n",
       "S:222  S:222\\nnegative regulation of protein cataboli...  \n",
       "S:223  S:223\\nK63-linked polyubiquitin modification-d...  \n",
       "...                                                  ...  \n",
       "S:238                                     S:238\\nXY body  \n",
       "S:239                                S:239\\nGINS complex  \n",
       "S:240                           S:240\\nFANCM-MHF complex  \n",
       "S:241                                 S:241\\ndeuterosome  \n",
       "S:242  S:242\\nDNA replication-dependent nucleosome as...  \n",
       "S:243  S:243\\ncrossover junction endodeoxyribonucleas...  \n",
       "S:244                  S:244\\nlipid biosynthetic process  \n",
       "S:245                        S:245\\nrecombinase activity  \n",
       "S:246  S:246\\nnuclear origin of replication recogniti...  \n",
       "S:247             S:247\\nFanconi anaemia nuclear complex  \n",
       "S:248  S:248\\nDNA damage response, signal transductio...  \n",
       "S:249       S:249\\nDNA replication preinitiation complex  \n",
       "S:250                      S:250\\nCtf18 RFC-like complex  \n",
       "S:251  S:251\\nresolution of meiotic recombination int...  \n",
       "S:252                           S:252\\ncondensin complex  \n",
       "S:253                         S:253\\nprotein-DNA complex  \n",
       "S:254  S:254\\ntelomere maintenance via semi-conservat...  \n",
       "S:255                             S:255\\nhistone binding  \n",
       "S:256                        S:256\\ncondensed chromosome  \n",
       "S:257               S:257\\nheterocyclic compound binding  \n",
       "S:258                               S:258\\ncell division  \n",
       "S:259                                  S:259\\nDNA repair  \n",
       "S:260    S:260\\ncellular response to DNA damage stimulus  \n",
       "S:261                                  S:261\\ncell cycle  \n",
       "S:262                S:262\\nintracellular organelle part  \n",
       "S:263                                     S:263\\nnucleus  \n",
       "S:264                             S:264\\nprotein binding  \n",
       "S:265                                     S:265\\nbinding  \n",
       "S:266                            S:266\\ncellular process  \n",
       "S:267                          S:267\\nintracellular part  \n",
       "\n",
       "[74 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note how node attributes have been updated to reflect the ontology alignment\n",
    "ont.node_attr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbpresent": {
     "id": "b508af71-63cd-4275-9b77-9e8338e1a345"
    }
   },
   "source": [
    "# Upload ontology with NDEx to visualize in the HiView application (http://hiview.ucsd.edu)\n",
    "* A two-dimensional layout of nodes is automatically calculated to optimize visualization of hierarchical structure\n",
    "* Molecular networks, such as protein-protein interactions and RNA coexpression, can be visualized in HiView to understand how an ontology's structure is consistent with data\n",
    "* Node attributes (color and size) can be set to visualize metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:28:09.045356Z",
     "start_time": "2018-12-07T22:28:09.037757Z"
    },
    "collapsed": true,
    "nbpresent": {
     "id": "bda55846-119b-4b2e-8d57-168b72b2193d"
    }
   },
   "outputs": [],
   "source": [
    "# Set the node color of seed genes to be green\n",
    "fill_attr = pd.DataFrame({'Vis:Fill Color' : '#6ACC65'}, index=seed)\n",
    "ont.update_node_attr(fill_attr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:28:09.055563Z",
     "start_time": "2018-12-07T22:28:09.047410Z"
    },
    "collapsed": true,
    "nbpresent": {
     "id": "dec08f16-ed17-4716-95cb-bdbf6e56e2d1"
    }
   },
   "outputs": [],
   "source": [
    "# Set the node color of inferred terms according to the alignment with GO (for visualization in HiView)\n",
    "fill_attr = ont.node_attr['Aligned_Similarity'].dropna().map(ddot.color_gradient)\n",
    "fill_attr = fill_attr.to_frame().rename(columns={'Aligned_Similarity' : 'Vis:Fill Color'})\n",
    "ont.update_node_attr(fill_attr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:28:25.913146Z",
     "start_time": "2018-12-07T22:28:09.057520Z"
    },
    "nbpresent": {
     "id": "44364aa3-b43a-485c-98e0-632ce5dfeb57"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene1</th>\n",
       "      <th>Gene2</th>\n",
       "      <th>PPI sum</th>\n",
       "      <th>Giant_5</th>\n",
       "      <th>GTEx Artery-Aorta</th>\n",
       "      <th>GTEx Breast-MammaryTissue</th>\n",
       "      <th>GTEx sum</th>\n",
       "      <th>PPI CCSB</th>\n",
       "      <th>GTEx Skin-SunExposed_Lowerleg</th>\n",
       "      <th>GTEx Adipose-Subcutaneous</th>\n",
       "      <th>...</th>\n",
       "      <th>Giant_4</th>\n",
       "      <th>PPI InbioMap</th>\n",
       "      <th>PPI BioGRID</th>\n",
       "      <th>PPI huMap</th>\n",
       "      <th>GTEx Skin-NotSunExposed_Suprapubic</th>\n",
       "      <th>GTEx Adipose-Visceral_Omentum</th>\n",
       "      <th>GTEx WholeBlood</th>\n",
       "      <th>GTEx Thyroid</th>\n",
       "      <th>GTEx Esophagus-Muscularis</th>\n",
       "      <th>RandomForest integrated similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BUB1B</td>\n",
       "      <td>CASP8AP2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5.472130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BUB1B</td>\n",
       "      <td>POLG2</td>\n",
       "      <td>0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>70.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6.042856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BUB1B</td>\n",
       "      <td>PLK4</td>\n",
       "      <td>0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>134.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>7.227561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BUB1B</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>105.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6.376164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BUB1B</td>\n",
       "      <td>NUDT21</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6.112902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene1     Gene2  PPI sum  Giant_5  GTEx Artery-Aorta  \\\n",
       "0  BUB1B  CASP8AP2        0      0.0              False   \n",
       "1  BUB1B     POLG2        0     14.0              False   \n",
       "2  BUB1B      PLK4        0    106.0              False   \n",
       "3  BUB1B    SRSF10        0     46.0              False   \n",
       "4  BUB1B    NUDT21        0      1.0              False   \n",
       "\n",
       "   GTEx Breast-MammaryTissue  GTEx sum  PPI CCSB  \\\n",
       "0                      False         0     False   \n",
       "1                      False         0     False   \n",
       "2                      False         3     False   \n",
       "3                      False         0     False   \n",
       "4                      False         0     False   \n",
       "\n",
       "   GTEx Skin-SunExposed_Lowerleg  GTEx Adipose-Subcutaneous  \\\n",
       "0                          False                      False   \n",
       "1                          False                      False   \n",
       "2                          False                      False   \n",
       "3                          False                      False   \n",
       "4                          False                      False   \n",
       "\n",
       "                  ...                  Giant_4  PPI InbioMap  PPI BioGRID  \\\n",
       "0                 ...                      0.0         False        False   \n",
       "1                 ...                     70.0         False        False   \n",
       "2                 ...                    134.0         False        False   \n",
       "3                 ...                    105.0         False        False   \n",
       "4                 ...                      4.0         False        False   \n",
       "\n",
       "   PPI huMap  GTEx Skin-NotSunExposed_Suprapubic  \\\n",
       "0      False                               False   \n",
       "1      False                               False   \n",
       "2      False                               False   \n",
       "3      False                               False   \n",
       "4      False                               False   \n",
       "\n",
       "   GTEx Adipose-Visceral_Omentum  GTEx WholeBlood  GTEx Thyroid  \\\n",
       "0                          False            False         False   \n",
       "1                          False            False         False   \n",
       "2                          False            False         False   \n",
       "3                          False            False         False   \n",
       "4                          False            False         False   \n",
       "\n",
       "   GTEx Esophagus-Muscularis  RandomForest integrated similarity  \n",
       "0                      False                            5.472130  \n",
       "1                      False                            6.042856  \n",
       "2                      False                            7.227561  \n",
       "3                      False                            6.376164  \n",
       "4                      False                            6.112902  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download a table containing multiple types of gene-gene interactions, which were preformatted and uploaded to NDEx for the Fanconi Anemia example.\n",
    "from ndex.networkn import NdexGraph\n",
    "G_ndex = NdexGraph(server=ddot.parse_ndex_server(ddot.FANGO_DATA_URL), uuid=ddot.parse_ndex_uuid(ddot.FANGO_DATA_URL))\n",
    "G = ddot.NdexGraph_to_nx(G_ndex)\n",
    "gene_network_data = ddot.nx_edges_to_pandas(G)\n",
    "gene_network_data.index.names = ['Gene1', 'Gene2']\n",
    "gene_network_data.reset_index(inplace=True)\n",
    "gene_network_data['RandomForest integrated similarity'] = [sim.loc[g1, g2] for g1, g2 in zip(gene_network_data['Gene1'], gene_network_data['Gene2'])]\n",
    "gene_network_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-07T22:34:49.692390Z",
     "start_time": "2018-12-07T22:34:00.839814Z"
    },
    "nbpresent": {
     "id": "341bae67-0f64-4c00-957b-3e49c8c135b0"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "To visualize in HiView, go to http://hiview.ucsd.edu in your web browser, and then\n",
      "\t--Enter this into the \"NDEx Sever URL\" field: http://dev2.ndexbio.org/\n",
      "\t--Enter this into the \"UUID of the main hierarchy\" field: 4e8f85b7-fa70-11e8-ad43-0660b7976219\n",
      "Alternatively, go to http://hiview.ucsd.edu/4e8f85b7-fa70-11e8-ad43-0660b7976219?type=test&server=http://dev2.ndexbio.org\n"
     ]
    }
   ],
   "source": [
    "# Upload ontology to NDEx\n",
    "ndex_url, G = ont.to_ndex(\n",
    "            name='Fanconi Anemia Gene Ontology',\n",
    "            description='Generated with the Data-driven Ontology Toolkit (https://github.com/michaelkyu/ddot)',\n",
    "            ndex_server=ndex_server,\n",
    "            ndex_user=ndex_user,\n",
    "            ndex_pass=ndex_pass,\n",
    "            visibility='PUBLIC',\n",
    "            layout='bubble-collect',\n",
    "            network=gene_network_data,\n",
    "            main_feature='RandomForest integrated similarity')\n",
    "\n",
    "print('To visualize in HiView, go to http://hiview.ucsd.edu in your web browser, and then')\n",
    "print('\\t--Enter this into the \"NDEx Sever URL\" field: %s' % ddot.parse_ndex_server(ndex_url))\n",
    "print('\\t--Enter this into the \"UUID of the main hierarchy\" field: %s' % ddot.parse_ndex_uuid(ndex_url))\n",
    "print('Alternatively, go to %s' % ddot.to_hiview_url(ndex_url))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "225px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_position": {
    "height": "690px",
    "left": "0px",
    "right": "1414.04px",
    "top": "108px",
    "width": "330px"
   },
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
